src/client/content_cao.cpp

```diff
@@ -1514,18 +1515,60 @@ void GenericCAO::updateAnimationSpeed()
 	m_animated_meshnode->setAnimationSpeed(m_animation_speed);
 }
 
-void GenericCAO::updateBonePosition()
+void GenericCAO::updateBonePosition(f32 dtime)
 {
-	if (m_bone_position.empty() || !m_animated_meshnode)
+	if (!m_animated_meshnode)
+		return;
+	if (m_bone_override.empty()) {
+		m_animated_meshnode->setJointMode(scene::EJUOR_NONE);
 		return;
+	}
 
 	m_animated_meshnode->setJointMode(scene::EJUOR_CONTROL); // To write positions to the mesh on render
-	for (auto &it : m_bone_position) {
+	for (auto &it : m_bone_override) {
 		std::string bone_name = it.first;
 		scene::IBoneSceneNode* bone = m_animated_meshnode->getJointNode(bone_name.c_str());
-		if (bone) {
-			bone->setPosition(it.second.X);
-			bone->setRotation(it.second.Y);
+		if (!bone)
+			continue;
+
+		BoneOverride *props = it.second;
+		props->dtime_passed += dtime;
+
+		// Position
+		{
+			v3f position = props->position.vector;
+			f32 progress = props->dtime_passed / props->position.interpolation_duration;
+			if (progress > 1.0f || props->position.interpolation_duration == 0.0f)
+				progress = 1.0f;
+			position = position.getInterpolated(props->position.previous, progress);
+			bone->setPosition(props->position.absolute
+					? position
+					: position + bone->getPosition());
+		}
+		// Rotation
+		{
+			core::quaternion rotation;
+			f32 progress = props->dtime_passed / props->rotation.interpolation_duration;
+			if (progress > 1.0f || props->rotation.interpolation_duration == 0.0f)
+				progress = 1.0f;
+			rotation.slerp(props->rotation.previous, props->rotation.next, progress);
+			if (!props->rotation.absolute) {
+				core::quaternion bone_rot(bone->getRotation() * core::DEGTORAD);
+				rotation = rotation * bone_rot; // first rotate around bone rot, then rot
+			}
+			v3f rot_euler;
+			rotation.toEuler(rot_euler);
+			bone->setRotation(rot_euler * core::RADTODEG);
+		}
+		// Scale
+		{
+			f32 progress = props->dtime_passed / props->scale.interpolation_duration;
+			if (progress > 1.0f || props->scale.interpolation_duration == 0.0f)
+				progress = 1.0f;
+			v3f scale = props->scale.vector.getInterpolated(props->scale.previous, progress);
+			bone->setScale(props->scale.absolute
+					? scale
+					: scale * bone->getScale());
```

Comment:
proper OOP _would_ be to have these implementations inside `BoneOverride`
(not saying you have to)